signac/contrib/job.py

```diff
@@ -483,6 +491,11 @@ def statepoint(self):
 
         return self._statepoint
 
+    # TODO: If we really believe what `reset_statepoint`'s documentation says,
+    # I think we should deprecate this functionality. If resetting a statepoint
+    # is a dangerous operation, making it settable via a property is too easy,
+    # and moreover it's impossible to document effectively because property
+    # setters don't show up in docs.
```

Comment:
> Just because something is potentially destructive does not mean it shouldn't exist.
Sure, but many APIs are designed to avoid providing users with foot guns unless they have demonstrable value. In this case I agree that setting a statepoint is a potentially useful operation, and I'm not advocating for making it impossible. However, it is inconsistent to mark `reset_statepoint` as "dangerous" while considering it obvious that `job.statepoint = ...` is destructive. I'm advocating for more consistency in the level of user protection between these two, and consequently making it slightly harder for a user to shoot themselves in the foot. By forcing them to call a method that they'd probably have to look up, we increase the chances that they see a documented warning.

I don't feel terribly strongly on this issue and would settle for updating the docstring of the `statepoint` property (it would have to be the getter since setter docstrings don't show up anywhere) to include the warning instead if you would rather leave the setter.

Reply:
I'd be in favor of a solution that reduces the number of equivalent ways to do things.

> [There should be one-- and preferably only one --obvious way to do it.](https://www.python.org/dev/peps/pep-0020/)

That would be one of the following:
- Deprecate (1.x) and remove (2.0) `reset_statepoint`, move the cautionary statements in the docstring to the `statepoint` property. I prefer this approach, even though it may reduce the likelihood of users finding the cautionary statements in the docs. Honestly, I still can't remember what `reset_statepoint` does from the function name alone. It's one of my least favorite API names, and I strictly use the `@statepoint.setter` in practice.
- Deprecate (1.x) and remove (2.0) `@statepoint.setter` and require the use of `reset_statepoint`.

Reply:
Exactly, that's where I'm coming from as well. We don't need both of these, and the way they're documented makes them seem different. I would be fine with keeping the setter instead of `reset_statepoint`, since I agree that it should be fairly obvious that this operation could impact the data space.